File: other/Solution5.java

package other;
public class Solution5 {
    public static String longestPalindrome(String s) {
        if(s.length() <= 1){
            return s;
        }
        String current = String.valueOf(s.charAt(0));
        int max_len = 1;
        for(int start = 1; start < s.length(); start++){
            if(start - (max_len + 1) / 2 >= 0 && start + (max_len + 1) / 2  + 1<= s.length()){
            String odd = s.substring(start - (max_len + 1) / 2, start +( max_len + 1)/ 2 + 1);
            if(isPalindrome(odd)){
                max_len = odd.length();
                current = odd;
                start--;
            }
        }
        if(start - (max_len + 2)/ 2  >= 0 && start + (max_len + 2) / 2  <= s.length()){
            String even = s.substring(start - (max_len + 2) / 2 , start + (max_len + 2)/ 2 );
            if(isPalindrome(even)){
                current = even;
                max_len = even.length();
                start--;
            }
        }
        }
        return current;
    }

    private static boolean isPalindrome(String s){
        if(s.length() < 2){
            return true;
        }
        char x = s.charAt(0);
        char y = s.charAt(s.length() - 1);
        if(x != y){
            return false;
        }
        return isPalindrome(s.substring(1, s.length() - 1));
    }

    public static void main(String[] args){
        String x = "ptmykvjanwiihepqhzupneckpzomgvzmyoybzfeggbgcillisyoskdodzlpbltefiz";
        System.out.println(longestPalindrome(x));
    }

}

File: C++/Solution5.cpp
#include<iostream>
#include<vector>

using namespace std;

class Solution5 {
public:
    string longestPalindrome(string s) {
        int start = 0;
        int end = 0;
        int maxlen = 1;
        int size = s.length();
        for(int center = 0; center < size; center++){
            int len = 1;
            while(center - len >= 0 && center + len < size && s[center - len] == s[center + len]){
                len += 1;
            }
            if(2 * len - 1 > maxlen){
                start = center - len + 1;
                end = center + len - 1;
                maxlen = end - start + 1;
            }
            int len2 = 0;
            while(center - len2 >= 0 && center + 1 + len2 < size && s[center - len2] == s[center + 1 + len2]){
                len2 += 1;
            }
            if(2 * len2 > maxlen){
                start = center - len2 + 1;
                end = center + len2;
                maxlen = 2 * len2;
            }
        }
         return s.substr(start, maxlen);
    }
};

int main(){
    Solution5 s;
    string a = "bb";
    cout << s.longestPalindrome(a);
    return 0;
}